File: windows-productivity/src/core/state.ts
/**
 * Watchtower Windows - State Management
 * Handles persistent state for tasks, sessions, and productivity tracking
 */

import { existsSync, readFileSync, writeFileSync, mkdirSync } from 'fs';
import { join, dirname } from 'path';
import dayjs from 'dayjs';
import { getConfig } from './config.js';
import type {
  Task,
  DailyEntry,
  EnergyReading,
  SprintStatus,
  Session,
  TaskPriority,
  EnergyLevel,
  PatternAnalysis,
  AvoidancePattern,
} from '../types.js';

// State file paths
function getStatePath(filename: string): string {
  const config = getConfig();
  return join(config.system.dataDir, filename);
}

/**
 * Generic state file operations
 */
function loadState<T>(filename: string, defaultValue: T): T {
  const path = getStatePath(filename);
  if (!existsSync(path)) {
    return defaultValue;
  }
  try {
    return JSON.parse(readFileSync(path, 'utf-8'));
  } catch {
    return defaultValue;
  }
}

function saveState<T>(filename: string, data: T): void {
  const path = getStatePath(filename);
  const dir = dirname(path);
  if (!existsSync(dir)) {
    mkdirSync(dir, { recursive: true });
  }
  writeFileSync(path, JSON.stringify(data, null, 2));
}

// ============================================
// TASK MANAGEMENT
// ============================================

interface TaskStore {
  tasks: Task[];
  lastUpdated: string;
}

export function loadTasks(): Task[] {
  const store = loadState<TaskStore>('tasks.json', { tasks: [], lastUpdated: '' });
  return store.tasks;
}

export function saveTasks(tasks: Task[]): void {
  saveState<TaskStore>('tasks.json', {
    tasks,
    lastUpdated: dayjs().toISOString(),
  });
}

export function addTask(content: string, priority: TaskPriority = 'standard'): Task {
  const tasks = loadTasks();
  const newTask: Task = {
    id: generateId(),
    content,
    priority,
    createdAt: dayjs().toISOString(),
    rollForwardCount: 0,
  };
  tasks.push(newTask);
  saveTasks(tasks);
  return newTask;
}

export function completeTask(taskId: string): Task | null {
  const tasks = loadTasks();
  const task = tasks.find((t) => t.id === taskId);
  if (task) {
    task.completedAt = dayjs().toISOString();
    saveTasks(tasks);
  }
  return task || null;
}

export function rollForwardTask(taskId: string): Task | null {
  const tasks = loadTasks();
  const task = tasks.find((t) => t.id === taskId);
  if (task) {
    task.rollForwardCount += 1;
    if (!task.notes) task.notes = [];
    task.notes.push(`Rolled forward on ${dayjs().format('YYYY-MM-DD')}`);
    saveTasks(tasks);
  }
  return task || null;
}

export function getTasksByPriority(priority: TaskPriority): Task[] {
  return loadTasks().filter((t) => t.priority === priority && !t.completedAt);
}

export function getAvoidedTasks(minRollCount: number = 3): Task[] {
  return loadTasks().filter((t) => t.rollForwardCount >= minRollCount && !t.completedAt);
}

// ============================================
// DAILY ENTRIES
// ============================================

interface DailyStore {
  entries: Record<string, DailyEntry>;
}

export function loadDailyEntries(): Record<string, DailyEntry> {
  const store = loadState<DailyStore>('daily.json', { entries: {} });
  return store.entries;
}

export function saveDailyEntries(entries: Record<string, DailyEntry>): void {
  saveState<DailyStore>('daily.json', { entries });
}

export function getTodayEntry(): DailyEntry {
  const today = dayjs().format('YYYY-MM-DD');
  const entries = loadDailyEntries();

  if (!entries[today]) {
    const sprint = getSprintStatus();
    entries[today] = {
      date: today,
      sprintDay: sprint.currentDay,
      energyReadings: [],
      tasksCompleted: [],
      tasksRolledForward: [],
      fieldReports: [],
    };
    saveDailyEntries(entries);
  }

  return entries[today];
}

export function updateTodayEntry(updates: Partial<DailyEntry>): DailyEntry {
  const today = dayjs().format('YYYY-MM-DD');
  const entries = loadDailyEntries();
  const current = getTodayEntry();

  entries[today] = { ...current, ...updates };
  saveDailyEntries(entries);

  return entries[today];
}

export function addFieldReport(report: string): void {
  const entry = getTodayEntry();
  entry.fieldReports.push(`[${dayjs().format('HH:mm')}] ${report}`);
  updateTodayEntry(entry);
}

// ============================================
// ENERGY TRACKING
// ============================================

export function logEnergy(level: EnergyLevel, context?: string): EnergyReading {
  const reading: EnergyReading = {
    timestamp: dayjs().toISOString(),
    level,
    context,
  };

  const entry = getTodayEntry();
  entry.energyReadings.push(reading);
  updateTodayEntry(entry);

  return reading;
}

export function getRecentEnergyReadings(days: number = 7): EnergyReading[] {
  const entries = loadDailyEntries();
  const cutoff = dayjs().subtract(days, 'day');
  const readings: EnergyReading[] = [];

  for (const [date, entry] of Object.entries(entries)) {
    if (dayjs(date).isAfter(cutoff)) {
      readings.push(...entry.energyReadings);
    }
  }

  return readings.sort((a, b) => dayjs(a.timestamp).unix() - dayjs(b.timestamp).unix());
}

export function calculateAverageEnergy(readings: EnergyReading[]): number {
  if (readings.length === 0) return 0;

  const levelValues: Record<EnergyLevel, number> = {
    high: 5,
    medium: 4,
    low: 3,
    depleted: 2,
    recovery: 1,
  };

  const sum = readings.reduce((acc, r) => acc + levelValues[r.level], 0);
  return sum / readings.length;
}

// ============================================
// SPRINT TRACKING
// ============================================

interface SprintStore {
  currentDay: number;
  startDate: string;
  lastWorkDay: string;
  restDays: string[];
}

export function getSprintStatus(): SprintStatus {
  const store = loadState<SprintStore>('sprint.json', {
    currentDay: 1,
    startDate: dayjs().format('YYYY-MM-DD'),
    lastWorkDay: dayjs().format('YYYY-MM-DD'),
    restDays: [],
  });

  const config = getConfig();
  const today = dayjs().format('YYYY-MM-DD');

  // Check if we need to increment the day
  if (store.lastWorkDay !== today) {
    const daysSinceLastWork = dayjs(today).diff(dayjs(store.lastWorkDay), 'day');

    if (daysSinceLastWork > 1) {
      // Rest day(s) detected - reset sprint
      store.currentDay = 1;
      store.startDate = today;
    } else {
      store.currentDay += 1;
    }

    store.lastWorkDay = today;
    saveState('sprint.json', store);
  }

  // Determine status
  let status: 'healthy' | 'warning' | 'danger' = 'healthy';
  if (store.currentDay >= config.sprint.dangerDay) {
    status = 'danger';
  } else if (store.currentDay >= config.sprint.warningDay) {
    status = 'warning';
  }

  return {
    currentDay: store.currentDay,
    startDate: store.startDate,
    status,
    lastRestDay: store.restDays[store.restDays.length - 1],
  };
}

export function recordRestDay(): void {
  const store = loadState<SprintStore>('sprint.json', {
    currentDay: 0,
    startDate: dayjs().format('YYYY-MM-DD'),
    lastWorkDay: dayjs().format('YYYY-MM-DD'),
    restDays: [],
  });

  store.restDays.push(dayjs().format('YYYY-MM-DD'));
  store.currentDay = 0;
  store.startDate = dayjs().add(1, 'day').format('YYYY-MM-DD');
  saveState('sprint.json', store);
}

// ============================================
// SESSION MANAGEMENT
// ============================================

interface SessionStore {
  currentSession?: Session;
  history: Session[];
}

export function getCurrentSession(): Session | null {
  const store = loadState<SessionStore>('sessions.json', { history: [] });
  return store.currentSession || null;
}

export function startSession(type: Session['type']): Session {
  const store = loadState<SessionStore>('sessions.json', { history: [] });

  // Archive current session if exists
  if (store.currentSession) {
    store.history.push(store.currentSession);
  }

  const session: Session = {
    id: generateId(),
    startedAt: dayjs().toISOString(),
    lastActivity: dayjs().toISOString(),
    type,
  };

  store.currentSession = session;
  saveState('sessions.json', store);

  return session;
}

export function updateSession(updates: Partial<Session>): void {
  const store = loadState<SessionStore>('sessions.json', { history: [] });

  if (store.currentSession) {
    store.currentSession = {
      ...store.currentSession,
      ...updates,
      lastActivity: dayjs().toISOString(),
    };
    saveState('sessions.json', store);
  }
}

export function endSession(): void {
  const store = loadState<SessionStore>('sessions.json', { history: [] });

  if (store.currentSession) {
    store.history.push(store.currentSession);
    store.currentSession = undefined;
    saveState('sessions.json', store);
  }
}

// ============================================
// PATTERN ANALYSIS
// ============================================

export function analyzePatterns(): PatternAnalysis {
  const tasks = loadTasks();
  const entries = loadDailyEntries();
  const sprint = getSprintStatus();

  // Avoidance patterns
  const avoidancePatterns: AvoidancePattern[] = tasks
    .filter((t) => t.rollForwardCount >= 3 && !t.completedAt)
    .map((t) => ({
      taskId: t.id,
      taskContent: t.content,
      rollCount: t.rollForwardCount,
      firstRolled: t.notes?.[0]?.replace('Rolled forward on ', '') || t.createdAt,
      category: t.priority,
    }));

  // Energy trends by time of day
  const allReadings = getRecentEnergyReadings(7);
  const morningReadings = allReadings.filter((r) => {
    const hour = dayjs(r.timestamp).hour();
    return hour >= 6 && hour < 12;
  });
  const afternoonReadings = allReadings.filter((r) => {
    const hour = dayjs(r.timestamp).hour();
    return hour >= 12 && hour < 18;
  });
  const eveningReadings = allReadings.filter((r) => {
    const hour = dayjs(r.timestamp).hour();
    return hour >= 18 || hour < 6;
  });

  const energyTrends = [
    {
      period: 'morning' as const,
      averageLevel: calculateAverageEnergy(morningReadings),
      sampleCount: morningReadings.length,
    },
    {
      period: 'afternoon' as const,
      averageLevel: calculateAverageEnergy(afternoonReadings),
      sampleCount: afternoonReadings.length,
    },
    {
      period: 'evening' as const,
      averageLevel: calculateAverageEnergy(eveningReadings),
      sampleCount: eveningReadings.length,
    },
  ];

  // Completion rate
  const recentDays = Object.values(entries)
    .filter((e) => dayjs(e.date).isAfter(dayjs().subtract(7, 'day')));
  const totalCompleted = recentDays.reduce((acc, e) => acc + e.tasksCompleted.length, 0);
  const totalRolled = recentDays.reduce((acc, e) => acc + e.tasksRolledForward.length, 0);
  const completionRate = totalCompleted + totalRolled > 0
    ? totalCompleted / (totalCompleted + totalRolled)
    : 0;

  // Category balance
  const activeTasks = tasks.filter((t) => !t.completedAt);
  const categoryBalance = {
    deep: activeTasks.filter((t) => t.priority === 'deep').length,
    standard: activeTasks.filter((t) => t.priority === 'standard').length,
    light: activeTasks.filter((t) => t.priority === 'light').length,
    someday: activeTasks.filter((t) => t.priority === 'someday').length,
  };

  // Burnout risk assessment
  let burnoutRisk: 'low' | 'medium' | 'high' = 'low';
  const avgEnergy = calculateAverageEnergy(allReadings);

  if (sprint.status === 'danger' || avgEnergy < 2.5) {
    burnoutRisk = 'high';
  } else if (sprint.status === 'warning' || avgEnergy < 3.5) {
    burnoutRisk = 'medium';
  }

  return {
    avoidancePatterns,
    energyTrends,
    completionRate,
    categoryBalance,
    burnoutRisk,
  };
}

// ============================================
// UTILITIES
// ============================================

function generateId(): string {
  return `${Date.now()}-${Math.random().toString(36).substring(2, 9)}`;
}

export function formatTaskList(tasks: Task[]): string {
  if (tasks.length === 0) return 'No tasks';

  return tasks
    .map((t) => {
      const rollIndicator = t.rollForwardCount > 0 ? ` [rolled ${t.rollForwardCount}x]` : '';
      const status = t.completedAt ? '[x]' : '[ ]';
      return `${status} ${t.content}${rollIndicator}`;
    })
    .join('\n');
}

export function getDataSummary(): Record<string, unknown> {
  const tasks = loadTasks();
  const entries = loadDailyEntries();
  const sprint = getSprintStatus();
  const patterns = analyzePatterns();

  return {
    totalTasks: tasks.length,
    activeTasks: tasks.filter((t) => !t.completedAt).length,
    completedTasks: tasks.filter((t) => t.completedAt).length,
    avoidedTasks: patterns.avoidancePatterns.length,
    dailyEntries: Object.keys(entries).length,
    sprintDay: sprint.currentDay,
    sprintStatus: sprint.status,
    burnoutRisk: patterns.burnoutRisk,
    completionRate: `${(patterns.completionRate * 100).toFixed(1)}%`,
  };
}

File: windows-productivity/src/watchtower/core/state.py
"""
Watchtower Windows - State Management
Handles persistent state for tasks, sessions, and productivity tracking.
"""

import json
import uuid
from datetime import datetime, timedelta
from pathlib import Path
from typing import Optional

from watchtower.core.config import get_config, get_data_dir
from watchtower.types import (
    Task,
    TaskPriority,
    DailyEntry,
    EnergyReading,
    EnergyLevel,
    SprintStatus,
    SprintStatusLevel,
    Session,
    SessionType,
    PatternAnalysis,
    AvoidancePattern,
    EnergyTrend,
    CategoryBalance,
    BurnoutRisk,
)


def generate_id() -> str:
    """Generate a unique ID."""
    return f"{int(datetime.now().timestamp())}-{uuid.uuid4().hex[:8]}"


def get_state_path(filename: str) -> Path:
    """Get path to a state file."""
    return get_data_dir() / filename


def load_json(filename: str, default: dict | list) -> dict | list:
    """Load JSON data from a state file."""
    path = get_state_path(filename)
    if not path.exists():
        return default
    try:
        with open(path, "r", encoding="utf-8") as f:
            return json.load(f)
    except (json.JSONDecodeError, IOError):
        return default


def save_json(filename: str, data: dict | list) -> None:
    """Save JSON data to a state file."""
    path = get_state_path(filename)
    path.parent.mkdir(parents=True, exist_ok=True)
    with open(path, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=2, default=str)


# ============================================
# TASK MANAGEMENT
# ============================================


def load_tasks() -> list[Task]:
    """Load all tasks from storage."""
    data = load_json("tasks.json", {"tasks": []})
    return [Task.model_validate(t) for t in data.get("tasks", [])]


def save_tasks(tasks: list[Task]) -> None:
    """Save tasks to storage."""
    save_json("tasks.json", {
        "tasks": [t.model_dump() for t in tasks],
        "last_updated": datetime.now().isoformat(),
    })


def add_task(
    content: str,
    priority: TaskPriority = TaskPriority.STANDARD,
    notes: Optional[list[str]] = None
) -> Task:
    """Add a new task."""
    tasks = load_tasks()
    new_task = Task(
        id=generate_id(),
        content=content,
        priority=priority,
        created_at=datetime.now(),
        notes=notes or [],
    )
    tasks.append(new_task)
    save_tasks(tasks)
    return new_task


def complete_task(task_id: str) -> Optional[Task]:
    """Mark a task as completed."""
    tasks = load_tasks()
    for task in tasks:
        if task.id == task_id:
            task.completed_at = datetime.now()
            save_tasks(tasks)
            return task
    return None


def roll_forward_task(task_id: str) -> Optional[Task]:
    """Roll forward an incomplete task."""
    tasks = load_tasks()
    for task in tasks:
        if task.id == task_id:
            task.roll_forward_count += 1
            task.notes.append(f"Rolled forward on {datetime.now().strftime('%Y-%m-%d')}")
            save_tasks(tasks)
            return task
    return None


def find_task(identifier: str) -> Optional[Task]:
    """Find a task by ID or content match."""
    tasks = load_tasks()

    # Try exact ID match first
    for task in tasks:
        if task.id == identifier:
            return task

    # Try content match
    identifier_lower = identifier.lower()
    for task in tasks:
        if identifier_lower in task.content.lower():
            return task

    return None


def get_tasks_by_priority(priority: TaskPriority) -> list[Task]:
    """Get active tasks by priority."""
    return [t for t in load_tasks() if t.priority == priority and t.completed_at is None]


def get_avoided_tasks(min_roll_count: int = 3) -> list[Task]:
    """Get tasks that have been rolled forward multiple times."""
    return [
        t for t in load_tasks()
        if t.roll_forward_count >= min_roll_count and t.completed_at is None
    ]


# ============================================
# DAILY ENTRIES
# ============================================


def load_daily_entries() -> dict[str, DailyEntry]:
    """Load all daily entries."""
    data = load_json("daily.json", {"entries": {}})
    entries = {}
    for date, entry_data in data.get("entries", {}).items():
        entries[date] = DailyEntry.model_validate(entry_data)
    return entries


def save_daily_entries(entries: dict[str, DailyEntry]) -> None:
    """Save daily entries."""
    save_json("daily.json", {
        "entries": {date: entry.model_dump() for date, entry in entries.items()}
    })


def get_today_entry() -> DailyEntry:
    """Get or create today's entry."""
    today = datetime.now().strftime("%Y-%m-%d")
    entries = load_daily_entries()

    if today not in entries:
        sprint = get_sprint_status()
        entries[today] = DailyEntry(
            date=today,
            sprint_day=sprint.current_day,
        )
        save_daily_entries(entries)

    return entries[today]


def update_today_entry(updates: dict) -> DailyEntry:
    """Update today's entry with new values."""
    today = datetime.now().strftime("%Y-%m-%d")
    entries = load_daily_entries()
    current = get_today_entry()

    # Merge updates
    entry_dict = current.model_dump()
    entry_dict.update(updates)
    entries[today] = DailyEntry.model_validate(entry_dict)
    save_daily_entries(entries)

    return entries[today]


def add_field_report(report: str) -> None:
    """Add a field report to today's entry."""
    entry = get_today_entry()
    timestamp = datetime.now().strftime("%H:%M")
    entry.field_reports.append(f"[{timestamp}] {report}")
    update_today_entry({"field_reports": entry.field_reports})


# ============================================
# ENERGY TRACKING
# ============================================


def log_energy(level: EnergyLevel, context: Optional[str] = None) -> EnergyReading:
    """Log an energy reading."""
    reading = EnergyReading(
        timestamp=datetime.now(),
        level=level,
        context=context,
    )

    entry = get_today_entry()
    entry.energy_readings.append(reading)
    update_today_entry({"energy_readings": [r.model_dump() for r in entry.energy_readings]})

    return reading


def get_recent_energy_readings(days: int = 7) -> list[EnergyReading]:
    """Get energy readings from the past N days."""
    entries = load_daily_entries()
    cutoff = datetime.now() - timedelta(days=days)
    readings: list[EnergyReading] = []

    for date, entry in entries.items():
        entry_date = datetime.strptime(date, "%Y-%m-%d")
        if entry_date >= cutoff:
            readings.extend(entry.energy_readings)

    return sorted(readings, key=lambda r: r.timestamp)


def calculate_average_energy(readings: list[EnergyReading]) -> float:
    """Calculate average energy level from readings."""
    if not readings:
        return 0.0

    level_values = {
        EnergyLevel.HIGH: 5,
        EnergyLevel.MEDIUM: 4,
        EnergyLevel.LOW: 3,
        EnergyLevel.DEPLETED: 2,
        EnergyLevel.RECOVERY: 1,
    }

    total = sum(level_values[r.level] for r in readings)
    return total / len(readings)


# ============================================
# SPRINT TRACKING
# ============================================


def get_sprint_status() -> SprintStatus:
    """Get current sprint status."""
    data = load_json("sprint.json", {
        "current_day": 1,
        "start_date": datetime.now().strftime("%Y-%m-%d"),
        "last_work_day": datetime.now().strftime("%Y-%m-%d"),
        "rest_days": [],
    })

    config = get_config()
    today = datetime.now().strftime("%Y-%m-%d")

    # Check if we need to update the day count
    if data.get("last_work_day") != today:
        last_work = datetime.strptime(data["last_work_day"], "%Y-%m-%d")
        today_date = datetime.strptime(today, "%Y-%m-%d")
        days_since = (today_date - last_work).days

        if days_since > 1:
            # Rest day(s) detected - reset sprint
            data["current_day"] = 1
            data["start_date"] = today
        else:
            data["current_day"] = data.get("current_day", 0) + 1

        data["last_work_day"] = today
        save_json("sprint.json", data)

    # Determine status
    current_day = data.get("current_day", 1)
    if current_day >= config.sprint.danger_day:
        status = SprintStatusLevel.DANGER
    elif current_day >= config.sprint.warning_day:
        status = SprintStatusLevel.WARNING
    else:
        status = SprintStatusLevel.HEALTHY

    rest_days = data.get("rest_days", [])

    return SprintStatus(
        current_day=current_day,
        start_date=data.get("start_date", today),
        status=status,
        last_rest_day=rest_days[-1] if rest_days else None,
    )


def record_rest_day() -> None:
    """Record a rest day and reset sprint counter."""
    data = load_json("sprint.json", {
        "current_day": 0,
        "start_date": datetime.now().strftime("%Y-%m-%d"),
        "last_work_day": datetime.now().strftime("%Y-%m-%d"),
        "rest_days": [],
    })

    today = datetime.now().strftime("%Y-%m-%d")
    tomorrow = (datetime.now() + timedelta(days=1)).strftime("%Y-%m-%d")

    rest_days = data.get("rest_days", [])
    rest_days.append(today)

    data["rest_days"] = rest_days
    data["current_day"] = 0
    data["start_date"] = tomorrow

    save_json("sprint.json", data)


# ============================================
# SESSION MANAGEMENT
# ============================================


def get_current_session() -> Optional[Session]:
    """Get the current active session."""
    data = load_json("sessions.json", {"history": []})
    current = data.get("current_session")
    if current:
        return Session.model_validate(current)
    return None


def start_session(session_type: SessionType) -> Session:
    """Start a new session."""
    data = load_json("sessions.json", {"history": []})

    # Archive current session if exists
    if data.get("current_session"):
        data["history"].append(data["current_session"])

    session = Session(
        id=generate_id(),
        started_at=datetime.now(),
        last_activity=datetime.now(),
        type=session_type,
    )

    data["current_session"] = session.model_dump()
    save_json("sessions.json", data)

    return session


def end_session() -> None:
    """End the current session."""
    data = load_json("sessions.json", {"history": []})

    if data.get("current_session"):
        data["history"].append(data["current_session"])
        data["current_session"] = None

    save_json("sessions.json", data)


# ============================================
# PATTERN ANALYSIS
# ============================================


def analyze_patterns() -> PatternAnalysis:
    """Analyze productivity patterns."""
    tasks = load_tasks()
    entries = load_daily_entries()
    sprint = get_sprint_status()
    config = get_config()

    # Avoidance patterns
    avoidance_patterns = [
        AvoidancePattern(
            task_id=t.id,
            task_content=t.content,
            roll_count=t.roll_forward_count,
            first_rolled=t.notes[0].replace("Rolled forward on ", "") if t.notes else t.created_at.isoformat(),
            category=t.priority,
        )
        for t in tasks
        if t.roll_forward_count >= 3 and t.completed_at is None
    ]

    # Energy trends by time of day
    all_readings = get_recent_energy_readings(7)

    morning_readings = [
        r for r in all_readings
        if 6 <= r.timestamp.hour < 12
    ]
    afternoon_readings = [
        r for r in all_readings
        if 12 <= r.timestamp.hour < 18
    ]
    evening_readings = [
        r for r in all_readings
        if r.timestamp.hour >= 18 or r.timestamp.hour < 6
    ]

    energy_trends = [
        EnergyTrend(
            period="morning",
            average_level=calculate_average_energy(morning_readings),
            sample_count=len(morning_readings),
        ),
        EnergyTrend(
            period="afternoon",
            average_level=calculate_average_energy(afternoon_readings),
            sample_count=len(afternoon_readings),
        ),
        EnergyTrend(
            period="evening",
            average_level=calculate_average_energy(evening_readings),
            sample_count=len(evening_readings),
        ),
    ]

    # Completion rate
    cutoff = datetime.now() - timedelta(days=7)
    recent_entries = [
        e for date, e in entries.items()
        if datetime.strptime(date, "%Y-%m-%d") >= cutoff
    ]
    total_completed = sum(len(e.tasks_completed) for e in recent_entries)
    total_rolled = sum(len(e.tasks_rolled_forward) for e in recent_entries)
    completion_rate = (
        total_completed / (total_completed + total_rolled)
        if (total_completed + total_rolled) > 0
        else 0.0
    )

    # Category balance
    active_tasks = [t for t in tasks if t.completed_at is None]
    category_balance = CategoryBalance(
        deep=len([t for t in active_tasks if t.priority == TaskPriority.DEEP]),
        standard=len([t for t in active_tasks if t.priority == TaskPriority.STANDARD]),
        light=len([t for t in active_tasks if t.priority == TaskPriority.LIGHT]),
        someday=len([t for t in active_tasks if t.priority == TaskPriority.SOMEDAY]),
    )

    # Burnout risk assessment
    avg_energy = calculate_average_energy(all_readings)

    if sprint.status == SprintStatusLevel.DANGER or avg_energy < 2.5:
        burnout_risk = BurnoutRisk.HIGH
    elif sprint.status == SprintStatusLevel.WARNING or avg_energy < 3.5:
        burnout_risk = BurnoutRisk.MEDIUM
    else:
        burnout_risk = BurnoutRisk.LOW

    return PatternAnalysis(
        avoidance_patterns=avoidance_patterns,
        energy_trends=energy_trends,
        completion_rate=completion_rate,
        category_balance=category_balance,
        burnout_risk=burnout_risk,
    )


def get_data_summary() -> dict:
    """Get a summary of all productivity data."""
    tasks = load_tasks()
    entries = load_daily_entries()
    sprint = get_sprint_status()
    patterns = analyze_patterns()

    return {
        "total_tasks": len(tasks),
        "active_tasks": len([t for t in tasks if t.completed_at is None]),
        "completed_tasks": len([t for t in tasks if t.completed_at is not None]),
        "avoided_tasks": len(patterns.avoidance_patterns),
        "daily_entries": len(entries),
        "sprint_day": sprint.current_day,
        "sprint_status": sprint.status.value,
        "burnout_risk": patterns.burnout_risk.value,
        "completion_rate": f"{patterns.completion_rate * 100:.1f}%",
    }
